add blog for offline batch inference and ingestion using OSI (#3888)
opensearch-project/project-website: _posts/2025-07-30-Offline-Batch-ML-Inference-and-Ingestion-for-1-Billion-Requests-using-OpenSearch-Ingestion.md

Real-time vector generation approaches often incur higher costs per inference and are constrained by lower rate limits, creating significant bottlenecks during the ingestion phase. As you attempt to vectorize massive document collections, these limitations result in extended ingestion times and increased operational costs. The challenge is further compounded by the complexity of orchestrating the entire pipeline---from initial document processing to vector generation and final indexing into OpenSearch.

To address these vector generation challenges and streamline the ingestion workflow, we're excited to introduce a powerful integration between OpenSearch Ingestion (OSI) and ML Commons. This integration enables seamless batch ML inference processing in offline mode, allowing you to generate vectors for large-scale document collections. By optimizing the vector generation phase of the ingestion pipeline, we're making it easier than ever to build and scale neural search applications efficiently.

## How does ML Commons integrate with OSI?

> We have added a new [`ml_inference` processor](https://docs.opensearch.org/latest/data-prepper/pipelines/configuration/processors/ml-inference/) to OSI in order to integrate with ML Commons for creating offline batch inference jobs. Since OpenSearch 2.17, ML Commons has provided the [Batch Predict API](https://docs.opensearch.org/latestml-commons-plugin/api/model-apis/batch-predict/), which performs inference on large datasets in offline, asynchronous mode using models deployed on external model servers such as Amazon Bedrock, Amazon SageMaker, Cohere, and OpenAI. Integrating ML Commons with OSI incorporates the Batch Predict API into OSI, enabling OSI pipelines to run batch inference jobs during ingestion. The following diagram shows an OSI pipeline that orchestrates multiple components to perform this process end to end.

Reviewer: ```suggestion
We have added a new [`ml_inference` processor](https://docs.opensearch.org/latest/data-prepper/pipelines/configuration/processors/ml-inference/) to OpenSearch Ingestion in order to integrate with ML Commons for creating offline batch inference jobs. Since OpenSearch 2.17, ML Commons has provided the [Batch Predict API](https://docs.opensearch.org/latestml-commons-plugin/api/model-apis/batch-predict/), which performs inference on large datasets in offline, asynchronous mode using models deployed on external model servers such as Amazon Bedrock, Amazon SageMaker, Cohere, and OpenAI. Integrating ML Commons with OpenSearch Ingestion incorporates the Batch Predict API into OpenSearch Ingestion, enabling OpenSearch Ingestion pipelines to run batch inference jobs during ingestion. The following diagram shows an OpenSearch Ingestion pipeline that orchestrates multiple components to perform this process end to end.
```

        # Type of inference operation:
        # - batch_predict: for batch processing
        # - predict: for real-time inference
        action_type: "batch_predict"
        
        # Remote ML model service provider (bedrock or sagemaker)
        service_name: "bedrock"
        
        # Unique identifier for the ML model
        model_id: "EzNlGZcBo9m_Jklj4T0j"
        
        # S3 path where batch inference results will be stored
        output_path: "s3://xunzh-offlinebatch/bedrock-multisource/output-multisource/"
        
        # AWS configuration settings
        aws:
            # AWS Region where the Lambda function is deployed
            region: "us-west-2"
            # IAM role ARN for Lambda function execution
            sts_role_arn: "arn:aws:iam::388303208821:role/Admin"
            
        # Conditional expression that determines when to trigger the processor
        # In this case, only process when bucket matches "xunzh-offlinebatch"
        ml_when: /bucket == "xunzh-offlinebatch"
```

## Ingestion performance improvements using the ml_inference processor

> The OSI `ml_inference` processor significantly enhances data ingestion performance for ML-enabled search. It's ideally suited for use cases requiring ML-model-generated data, including semantic search, multimodal search, document enrichment, and query understanding. In semantic search, the processor can accelerate the creation and ingestion of large-volume, high-dimensional vectors by an order of magnitude.

Reviewer: ```suggestion
The OpenSearch Ingestion `ml_inference` processor significantly enhances data ingestion performance for ML-enabled search. It's ideally suited for use cases requiring ML-model-generated data, including semantic search, multimodal search, document enrichment, and query understanding. In semantic search, the processor can accelerate the creation and ingestion of large-volume, high-dimensional vectors by an order of magnitude.
```

1. In the OpenSearch Service console, go to **Integrations** and search for `SageMaker`.
1. Select **Integration with text embedding models through Amazon SageMaker**.
1. Choose **Configure domain**, then select **Configure VPC domain** or **Configure public domain** as appropriate, as shown in the following image.
    ![CloudFormation Console Integration](/assets/media/blog-images/2025-07-30-Offline-Batch-ML-Inference-and-Ingestion-for-1-Billion-Requests-using-OpenSearch-Ingestion/CFN_Console_Integration.png){: .img-fluid}
1. When starting the stack deployment, set the **Enable Offline Batch Inference** dropdown list to `true` to provision resources for offline batch processing, as shown in the following image.
    ![CloudFormation Parameters](/assets/media/blog-images/2025-07-30-Offline-Batch-ML-Inference-and-Ingestion-for-1-Billion-Requests-using-OpenSearch-Ingestion/CFN_Parameters.png){: .img-fluid}
1. After the CloudFormation stack is created, open the **Outputs** tab in the CloudFormation console to find the `connector_id` and `model_id`. You will need these values for later pipeline configuration steps.

> ### Step 2: Create an OSI pipeline for ingestion

Reviewer: ```suggestion
### Step 2: Create an OpenSearch Ingestion pipeline
```

> Create your OSI pipeline using the following code in the configuration editor:

Reviewer: ```suggestion
Create your OpenSearch Ingestion pipeline using the following code in the configuration editor:
```
